### ISSUE TITLE
p3d_LightSourceParameters shadowMap not universal.

### ISSUE THREAD

**Issue by lettier** [panda3d/panda3d#707]
Hello,

The [documentation](https://www.panda3d.org/manual/?title=List_of_GLSL_Shader_Inputs) states that the shadow map in `p3d_LightSourceParameters` is a `sampler2DShadow`. 

However, it is a [cube map](https://github.com/panda3d/panda3d/blob/daa57733cb9b4ccdb23e28153585e8e20b5ccdb5/panda/src/display/graphicsStateGuardian.cxx#L3304) for point lights. 

When using a custom GLSL shader, that uses the shadow map for both point and non-point lights, the following [error](https://github.com/panda3d/panda3d/blob/bc563c86643df9fc228c493c81d58a71d7f5fda3/panda/src/glstuff/glShaderContext_src.cxx#L2817) is continuously reported. 

```c
:display:gsg:glgsg(error): Sampler type of GLSL shader input p3d_LightSource[...].shadowMap does not match type of texture dummy-shadow-cube.
````

If you do set the point light as a shadow caster, the shadow map is still a [cube map](https://github.com/panda3d/panda3d/blob/daa57733cb9b4ccdb23e28153585e8e20b5ccdb5/panda/src/display/graphicsStateGuardian.cxx#L3338).

Code such as this [example](https://discourse.panda3d.org/t/shadows-direction-parameters-and-inputs-in-glsl/15300/4) is not universal across all light types. Using it with a point light would give you the error. 

Other than the shadow map, `p3d_LightSourceParameters` was made to be universal. For example, the `w` component being zero for directional lights or the spotlight cutoff being [-1](https://github.com/panda3d/panda3d/blob/daa57733cb9b4ccdb23e28153585e8e20b5ccdb5/panda/src/display/graphicsStateGuardian.cxx#L1705) for non-spotlights. 

Can a special struct be made just for point lights? I understand the need for six views given that a point light has no direction. Otherwise, can a dummy 2D texture be used for point lights with the expectation that point lights produce no shadows as per the [documentation](https://www.panda3d.org/manual/?title=Lighting#Shadow_Mapping)?

As of right now, you must either:

1) Use only point lights if you want shadows (doesn't seem to work).
2) Use point, directional, and spot but without shadows.
3) Use only directional and spot if you want shadows.

:+1:

**Comment by rdb** [comment]
Yes, you're right, this is unfortunate because it makes it impossible to create a shader that does both point light and other lights using p3d_LightSource.  A workaround is to pass in the point light explicitly using a shader input rather than via the usual light assignment system.

For the record, point light shadows *are* supported in GLSL (created [an issue](https://github.com/panda3d/panda3d-docs/issues/6) to clarify this in the manual) but they require a `samplerCubeShadow`, and the way to sample them is slightly different.  Let me know if I should dig up some sample code for point light GLSL shadows, I know I have it somewhere.

I think the way to fix this is to either (1) deprecate p3d_LightSource and instead adopt separate arrays for shadowed point lights, or (2) keep p3d_LightSource and put the shadow sources in separate arrays that are indexed into by the p3d_LightSource.

I guess with (1), we'd ideally have arrays like `p3d_UnshadowedLightSource`, `p3d_ShadowedPointLightSource`, `p3d_ShadowedNonPointLightSource`?  Not exactly pretty but that's a secondary concern.  Though, once we get support for IES profile lights, I fear that the amount of arrays would explode combinatorically.

The advantage of (2) is that you can sample only the shadow sources from their respective shadow map and therefore will perform better if only a few of your lights are shadowing, since it's easier to avoid having to do a dummy lookup for the other shadow maps.  However, with this option, we'd still need to invent some way to differentiate cube shadow maps from 2D shadow maps, and some branching would probably be needed for this.

**Comment by lettier** [comment]
You could keep them side-by-side like the following.

```c
uniform struct p3d_LightSourceParameters {
  sampler2DShadow shadowMap;
  samplerCubeShadow shadowMapCube;
} p3d_LightSource[...];
```

For those that only wish to use directional or spot, they only need to list `shadowMap`.

For those that only wish to use point, they only need to list `shadowMapCube`.

For those that need point and non-point, they would list the two and perform two texture look ups.

If the current `i` light is a non-point, `shadowMapCube` is set to `GL_NEVER` (or depth 1) resulting in no shadow. Similarly, if the current `i` light is a point, `shadowMap` is set to `GL_NEVER` (or depth 1) resulting in no shadow.

This method would also be backwards compatible, allowing shaders that only reference `shadowMap` to keep working.

:+1:

**Comment by rdb** [comment]
That's not a bad idea.  I'm not sure whether a branch might be better than two texture lookups, but we can offer both options and leave that freedom to the shader writer.

**Comment by lettier** [comment]
Yes I'm not sure which would perform better but that could be left up to the shader. I could try adding the `samplerCubeShadow` alongside the `sampler2DShadow` if you would like.

>  Let me know if I should dig up some sample code for point light GLSL shadows, I know I have it somewhere.

Yes, please, that would be excellent.

Looking at the [shader generator](https://github.com/panda3d/panda3d/blob/daa57733cb9b4ccdb23e28153585e8e20b5ccdb5/panda/src/pgraphnodes/shaderGenerator.cxx#L1344), it looks like it is:

1) Finding the maximum dimension of the vertex/fragment position in light/shadow space.
2) [Converting](https://en.wikipedia.org/wiki/Z-buffering#Mathematics) the maximum dimension or "Z" to Z' in the range [-1, 1].
3) Uses the vertex/fragment position in light/shadow space as the texture lookup direction. 
4) Converts Z' to [0,1] and compares it to the texture value.

I'm not sure what the `w` component is of the `latten` or attenuation vector is. I do know it is being used as the far plane. I guessing the near plane `lpoint.w` is one?

Here is what I have so far.

```c
vec4 vertexInShadowSpacesAbs = abs(vertexInShadowSpaces[i]);

float z =
  max
    ( vertexInShadowSpacesAbs.x
    , max
        ( vertexInShadowSpacesAbs.y
        , vertexInShadowSpacesAbs.z
        )
    );

float lightDistance = length(lightDirection);

float attenuation =
    1
  / ( p3d_LightSource[i].constantAttenuation
    + p3d_LightSource[i].linearAttenuation
    * lightDistance
    + p3d_LightSource[i].quadraticAttenuation
    * (lightDistance * lightDistance)
    );

float zPrime =
    ( (attenuation + p3d_LightSource[i].position.w)
    / (attenuation - p3d_LightSource[i].position.w)
    )
  +
    (1 / z)
  *
    ( (-2 * attenuation * p3d_LightSource[i].position.w)
    / (     attenuation - p3d_LightSource[i].position.w)
    );

float shadow =
  texture
    ( p3d_LightSource[i].shadowMap
    , vec4(vertexInShadowSpaces[i].xyz, zPrime)
    );

diffuse.rgb *= shadow;
```

However, this doesn't work unfortunately.

:+1:

**Comment by rdb** [comment]
I found a shader I hacked up during a very brief game jam, which works but it may not be the best way to do it.

Common:
```
uniform struct {
  vec4 position;
  vec3 color;
  vec3 attenuation;
  samplerCube shadowMap;
  mat4 shadowViewMatrix;
} lantern;
```

Vertex shader:
```
  vpos = vec3(p3d_ModelViewMatrix * vertex);
  shad = lantern.shadowViewMatrix * vec4(vpos, 1);
```

Fragment shader: 
```
in vec4 shad;
...
  vec4 coords;
  coords.xyz = shad.xyz / shad.w;
  float dist = max(abs(coords.x), max(abs(coords.y), abs(coords.z)));
  coords.w = (NEAR + FAR) / (FAR - NEAR) + ((-2 * FAR * NEAR) / (dist * (FAR - NEAR)));
  coords.w = coords.w * 0.5 + 0.5 + 0.0000001;

  float shadowMul = (texture(lantern.shadowMap, coords.xyz).r > coords.w - 0.001) ? 1.0 : 0.0;
```

**Comment by Moguri** [comment]
@rdb is @lettier's solution to add a `shadowMapCube` alongside `shadowMap` still on okay solution? If so, I may be interested in tackling this. I am not sure if this would be a 1.10 thing or 1.11 (i.e., a bug fix or a new feature). It adds something to the API, but shadows in custom shaders are kind of broken without them (cannot mix point lights with other lights without getting warnings or errors).

**Comment by rdb** [comment]
It seems fine to me.  It allows people to write shaders that are agnostic of light type and doesn't lock us out from alternative solutions in the future.

I think this would be a 1.11 thing.  We may be able to introduce this in a way that doesn't break backward compatibility, but there would be a risk of introducing bugs to the shadow system.

The workaround for 1.10 is to pass the p3d_LightSource for the point light using a shader input instead of the NodePath.set_light interface.
